File: Sources/FWObjC/Kernel/Task.m
//
//  Task.m
//  FWFramework
//
//  Created by wuyong on 2022/8/22.
//

#import "Task.h"

#pragma mark - __FWTask

typedef NS_ENUM(NSInteger, __FWTaskState) {
    __FWTaskStateCreated,
    __FWTaskStateReady = 1,
    __FWTaskStateLoading,
    __FWTaskStateSuccess,
    __FWTaskStateFailure,
    __FWTaskStateCanceled,
};

@interface __FWTask ()

@property (nonatomic, assign) __FWTaskState state;
@property (nonatomic, strong, readonly) NSRecursiveLock *lock;

@end

@implementation __FWTask

@synthesize lock = _lock;

- (instancetype)init
{
    self = [super init];
    if (self) {
        self.state = __FWTaskStateReady;
    }
    return self;
}

- (NSRecursiveLock *)lock
{
    if (!_lock) {
        _lock = [[NSRecursiveLock alloc] init];
    }
    return _lock;
}

- (void)start
{
    [self.lock lock];
    if ([self isReady]) {
        self.state = __FWTaskStateLoading;
        
        // 调试日志
        NSLog(@"\n********** TASK %@ STARTED", NSStringFromClass(self.class));
        
        [self.lock unlock];
        
        if ([self needMainThread]) {
            if ([NSThread isMainThread]) {
                [self executeTask];
            } else {
                [self performSelectorOnMainThread:@selector(executeTask) withObject:nil waitUntilDone:NO];
            }
        } else {
            [self executeTask];
        }
    } else {
        [self.lock unlock];
    }
}

- (void)executeTask
{
    @throw [NSException exceptionWithName:@"FWTask"
                                   reason:[NSString stringWithFormat:@"task %@ must override executeTask", [self.class description]]
                                 userInfo:nil];
}

- (void)finishWithError:(NSError *)error
{
    [self.lock lock];
    if (![self isFinished]) {
        if (error) {
            _error = error;
            self.state = __FWTaskStateFailure;
            
            // 调试日志
            NSLog(@"\n********** TASK %@ FAILED", NSStringFromClass(self.class));
        } else {
            self.state = __FWTaskStateSuccess;
            
            // 调试日志
            NSLog(@"\n********** TASK %@ FINISHED", NSStringFromClass(self.class));
        }
    }
    [self.lock unlock];
}

- (BOOL)needMainThread
{
    return NO;
}

- (void)cancel
{
    [self.lock lock];
    
    if (![self isFinished]) {
        self.state = __FWTaskStateCanceled;
        [super cancel];
        
        // 调试日志
        NSLog(@"\n********** TASK %@ CANCELLED", NSStringFromClass(self.class));
    }
    
    [self.lock unlock];
}

- (BOOL)isAsynchronous
{
    return YES;
}

- (BOOL)isReady
{
    return self.state == __FWTaskStateReady && [super isReady];
}

- (BOOL)isFinished
{
    return self.state == __FWTaskStateSuccess || self.state == __FWTaskStateFailure || self.state == __FWTaskStateCanceled;
}

- (BOOL)isExecuting
{
    return self.state == __FWTaskStateLoading;
}

- (BOOL)isValidTransition:(__FWTaskState)fromState toState:(__FWTaskState)toState
{
    switch (fromState) {
        case __FWTaskStateReady:
        {
            switch (toState) {
                case __FWTaskStateLoading:
                case __FWTaskStateSuccess:
                case __FWTaskStateFailure:
                case __FWTaskStateCanceled:
                    return YES;
                    break;
                default:
                    return NO;
                    break;
            }
            break;
        }
        case __FWTaskStateLoading:
        {
            switch (toState) {
                case __FWTaskStateSuccess:
                case __FWTaskStateFailure:
                case __FWTaskStateCanceled:
                    return YES;
                    break;
                default:
                    return NO;
                    break;
            }
        }
        case (__FWTaskState)0:
        {
            if (toState == __FWTaskStateReady) {
                return YES;
            } else {
                return NO;
            }
        }
        default:
            return NO;
            break;
    }
}

- (void)setState:(__FWTaskState)state
{
    [self.lock lock];
    if (![self isValidTransition:_state toState:state]) {
        [self.lock unlock];
        return;
    }
    
    switch (state) {
        case __FWTaskStateCanceled:
        {
            [self willChangeValueForKey:@"isExecuting"];
            [self willChangeValueForKey:@"isFinished"];
            [self willChangeValueForKey:@"isCancelled"];
            _state = state;
            [self didChangeValueForKey:@"isExecuting"];
            [self didChangeValueForKey:@"isFinished"];
            [self didChangeValueForKey:@"isCancelled"];
            break;
        }
        case __FWTaskStateLoading:
        {
            [self willChangeValueForKey:@"isExecuting"];
            _state = state;
            [self didChangeValueForKey:@"isExecuting"];
            break;
        }
        case __FWTaskStateSuccess:
        case __FWTaskStateFailure:
        {
            [self willChangeValueForKey:@"isFinished"];
            [self willChangeValueForKey:@"isExecuting"];
            _state = state;
            [self didChangeValueForKey:@"isFinished"];
            [self didChangeValueForKey:@"isExecuting"];
            break;
        }
        case __FWTaskStateReady:
        {
            [self willChangeValueForKey:@"isReady"];
            _state = state;
            [self didChangeValueForKey:@"isReady"];
            break;
        }
        default:
        {
            _state = state;
            break;
        }
    }
    
    [self.lock unlock];
}

@end

#pragma mark - __FWTaskManager

@implementation __FWTaskManager
{
    NSOperationQueue *_taskQueue;
}

+ (__FWTaskManager *)sharedInstance
{
    static __FWTaskManager *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[__FWTaskManager alloc] init];
    });
    return instance;
}

- (id)init
{
    self = [super init];
    if (self) {
        _taskQueue = [[NSOperationQueue alloc] init];
        _taskQueue.name = @"FWTaskManager.taskQueue";
    }
    return self;
}

- (NSInteger)maxConcurrentTaskCount
{
    return _taskQueue.maxConcurrentOperationCount;
}

- (void)setMaxConcurrentTaskCount:(NSInteger)maxConcurrentTaskCount
{
    _taskQueue.maxConcurrentOperationCount = maxConcurrentTaskCount;
}

- (BOOL)isSuspended
{
    return _taskQueue.isSuspended;
}

- (void)setIsSuspended:(BOOL)isSuspended
{
    _taskQueue.suspended = isSuspended;
}

- (void)addTask:(NSOperation *)task
{
    [_taskQueue addOperation:task];
}

- (void)addTasks:(NSArray<NSOperation *> *)tasks
{
    if (tasks.count > 0) {
        [_taskQueue addOperations:tasks waitUntilFinished:NO];
    }
}

- (void)addTaskConfig:(NSArray<NSDictionary *> *)config
{
    NSMutableDictionary *taskMap = [[NSMutableDictionary alloc] init];
    
    for (NSDictionary *taskInfo in config) {
        // className
        NSString *className = [taskInfo objectForKey:@"className"];
        Class clazz = NSClassFromString(className);
        if ([clazz isSubclassOfClass:[NSOperation class]]) {
            NSOperation *task = [[clazz alloc] init];
            [taskMap setObject:task forKey:className];
            // dependency
            NSArray *dependencyList = [[taskInfo objectForKey:@"dependency"] componentsSeparatedByString:@","];
            if (dependencyList.count) {
                for (NSString *depedencyClass in dependencyList) {
                    NSOperation *preTask = [taskMap objectForKey:depedencyClass];
                    if (preTask) [task addDependency:preTask];
                }
            }
        }
    }
    
    [self addTasks:[taskMap allValues]];
}

- (void)cancelAllTasks
{
    [_taskQueue cancelAllOperations];
}

- (void)waitUntilFinished
{
    [_taskQueue waitUntilAllOperationsAreFinished];
}

@end

File: Sources/FWFramework/Kernel/Task.swift
//
//  Task.swift
//  FWFramework
//
//  Created by wuyong on 2022/8/18.
//

import Foundation

// MARK: - Wrapper+DispatchQueue
extension Wrapper where Base: DispatchQueue {
    /// 主线程安全异步执行句柄
    public static func mainAsync(execute block: @escaping @MainActor @Sendable () -> Void) {
        MainActor.runAsync(execute: block)
    }

    /// 主线程安全同步执行句柄
    public static func mainSync(execute block: @MainActor () -> Void) {
        MainActor.runSync(execute: block)
    }

    /// 当主线程时执行句柄，非主线程不执行
    public static func mainSyncIf(execute block: @MainActor () -> Void) {
        MainActor.runSyncIf(execute: block)
    }
    
    /// 当主线程时执行句柄，非主线程执行另一个句柄
    public static func mainSyncIf(execute block: @MainActor () -> Void, otherwise: () -> Void) {
        MainActor.runSyncIf(execute: block, otherwise: otherwise)
    }
}

// MARK: - MainActor+Task
extension MainActor {
    /// 主Actor安全异步执行句柄
    public static func runAsync(execute block: @escaping @MainActor @Sendable () -> Void) {
        if Thread.isMainThread {
            MainActor.assumeIsolated(block)
        } else {
            DispatchQueue.main.async(execute: block)
        }
    }

    /// 主Actor安全同步执行句柄
    public static func runSync(execute block: @MainActor () -> Void) {
        if Thread.isMainThread {
            MainActor.assumeIsolated(block)
        } else {
            DispatchQueue.main.sync(execute: block)
        }
    }

    /// 当主线程时执行句柄，非主线程不执行
    public static func runSyncIf(execute block: @MainActor () -> Void) {
        if Thread.isMainThread {
            MainActor.assumeIsolated(block)
        }
    }
    
    /// 当主线程时执行句柄，非主线程另一个句柄
    public static func runSyncIf(execute block: @MainActor () -> Void, otherwise: () -> Void) {
        if Thread.isMainThread {
            MainActor.assumeIsolated(block)
        } else {
            otherwise()
        }
    }
}

// MARK: - TaskOperation
/// 任务操作类，可继承或直接使用
open class TaskOperation: Operation, @unchecked Sendable {
    private enum TaskState: Int {
        case created = 0
        case ready
        case loading
        case success
        case failure
        case cancelled
    }

    /// 任务句柄，执行完成需调用task.finish(error:)
    open var taskBlock: ((TaskOperation) -> Void)?

    /// 是否在主线程执行，会阻碍UI渲染，默认false
    open var onMainThread = false

    /// 任务错误信息
    open private(set) var error: Error?

    override public init() {
        super.init()
        self.state = .ready
    }

    public convenience init(onMainThread: Bool = false, queuePriority: Operation.QueuePriority = .normal, taskBlock: ((TaskOperation) -> Void)?) {
        self.init()
        self.onMainThread = onMainThread
        self.queuePriority = queuePriority
        self.taskBlock = taskBlock
    }

    /// 子类可重写，默认调用taskBlock，任务完成需调用finish(error:)
    @objc open func executeTask() {
        taskBlock?(self)
    }

    /// 是否主线程执行，子类可重写，会阻碍UI渲染，默认返回onMainThread
    open func needMainThread() -> Bool {
        onMainThread
    }

    /// 标记任务完成，error为空表示任务成功
    open func finish(error: Error? = nil) {
        lock.lock()
        guard !isFinished else {
            lock.unlock()
            return
        }

        if error != nil {
            self.error = error
            state = .failure

            #if DEBUG
            Logger.debug(group: Logger.fw.moduleName, "Task failed: %@", NSStringFromClass(type(of: self)))
            #endif
        } else {
            state = .success

            #if DEBUG
            Logger.debug(group: Logger.fw.moduleName, "Task finished: %@", NSStringFromClass(type(of: self)))
            #endif
        }
        lock.unlock()
    }

    override open func start() {
        lock.lock()
        guard isReady else {
            lock.unlock()
            return
        }

        state = .loading
        lock.unlock()

        #if DEBUG
        Logger.debug(group: Logger.fw.moduleName, "Task started: %@", NSStringFromClass(type(of: self)))
        #endif

        if needMainThread() {
            if Thread.isMainThread {
                executeTask()
            } else {
                performSelector(onMainThread: #selector(executeTask), with: nil, waitUntilDone: false)
            }
        } else {
            executeTask()
        }
    }

    override open func cancel() {
        lock.lock()

        if !isFinished {
            state = .cancelled
            super.cancel()

            #if DEBUG
            Logger.debug(group: Logger.fw.moduleName, "Task cancelled: %@", NSStringFromClass(type(of: self)))
            #endif
        }

        lock.unlock()
    }

    override open var isAsynchronous: Bool {
        true
    }

    override open var isReady: Bool {
        state == .ready && super.isReady
    }

    override open var isFinished: Bool {
        state == .success || state == .failure || state == .cancelled
    }

    override open var isExecuting: Bool {
        state == .loading
    }

    private var state: TaskState {
        get {
            _state
        }
        set {
            lock.lock()
            if !isValidTransition(from: _state, to: newValue) {
                lock.unlock()
                return
            }

            switch newValue {
            case .cancelled:
                willChangeValue(forKey: "isExecuting")
                willChangeValue(forKey: "isFinished")
                willChangeValue(forKey: "isCancelled")
                _state = newValue
                didChangeValue(forKey: "isExecuting")
                didChangeValue(forKey: "isFinished")
                didChangeValue(forKey: "isCancelled")
            case .loading:
                willChangeValue(forKey: "isExecuting")
                _state = newValue
                didChangeValue(forKey: "isExecuting")
            case .success, .failure:
                willChangeValue(forKey: "isFinished")
                willChangeValue(forKey: "isExecuting")
                _state = newValue
                didChangeValue(forKey: "isFinished")
                didChangeValue(forKey: "isExecuting")
            case .ready:
                willChangeValue(forKey: "isReady")
                _state = newValue
                didChangeValue(forKey: "isReady")
            default:
                _state = newValue
            }
            lock.unlock()
        }
    }

    private var _state: TaskState = .created

    private var lock = NSRecursiveLock()

    private func isValidTransition(from fromState: TaskState, to toState: TaskState) -> Bool {
        switch fromState {
        case .ready:
            switch toState {
            case .loading:
                return true
            case .success:
                return true
            case .failure:
                return true
            case .cancelled:
                return true
            default:
                return false
            }
        case .loading:
            switch toState {
            case .success:
                return true
            case .failure:
                return true
            case .cancelled:
                return true
            default:
                return false
            }
        case .created:
            if toState == .ready {
                return true
            } else {
                return false
            }
        default:
            return false
        }
    }
}

// MARK: - TaskManager
/// 任务管理器，兼容NSBlockOperation和NSInvocationOperation
open class TaskManager: @unchecked Sendable {
    /// 单例模式
    public static let shared = TaskManager()

    /// 并发操作的最大任务数
    open var maxConcurrentTaskCount: Int {
        get { taskQueue.maxConcurrentOperationCount }
        set { taskQueue.maxConcurrentOperationCount = newValue }
    }

    /// 是否暂停，可恢复
    open var isSuspended: Bool {
        get { taskQueue.isSuspended }
        set { taskQueue.isSuspended = newValue }
    }

    private lazy var taskQueue: OperationQueue = {
        let queue = OperationQueue()
        queue.name = "FWTaskManager.taskQueue"
        return queue
    }()

    public init() {}

    public convenience init(maxConcurrentTaskCount: Int, isSuspended: Bool = false) {
        self.init()
        self.maxConcurrentTaskCount = maxConcurrentTaskCount
        if isSuspended { self.isSuspended = isSuspended }
    }

    /// 添加单个任务
    open func addTask(_ task: Operation) {
        taskQueue.addOperation(task)
    }

    /// 批量添加任务
    open func addTasks(_ tasks: [Operation]) {
        guard !tasks.isEmpty else { return }
        taskQueue.addOperations(tasks, waitUntilFinished: false)
    }

    /// 从配置数组按顺序添加任务，支持className|dependency
    open func addTaskConfig(_ config: [[String: String]]) {
        var tasks: [Operation] = []
        var taskMap: [String: Operation] = [:]

        for taskInfo in config {
            // className
            if let className = taskInfo["className"],
               let operationClass = NSClassFromString(className) as? Operation.Type {
                let task = operationClass.init()
                tasks.append(task)
                taskMap[className] = task

                // dependency
                if let dependencyList = taskInfo["dependency"]?.components(separatedBy: ","),
                   !dependencyList.isEmpty {
                    for dependencyClass in dependencyList {
                        if let preTask = taskMap[dependencyClass] {
                            task.addDependency(preTask)
                        }
                    }
                }
            }
        }

        addTasks(tasks)
    }

    /// 取消所有任务
    open func cancelAllTasks() {
        taskQueue.cancelAllOperations()
    }

    /// 等待所有任务执行完成，会阻塞线程
    open func waitUntilFinished() {
        taskQueue.waitUntilAllOperationsAreFinished()
    }
}
